jexl-eval/src/lib.rs

```diff
@@ -67,103 +71,214 @@ impl Truthy for Value {
-    let context = value!({});
-    eval_in_context(input, &context)
+/// TransformFn represents an arbitrary transform function
+/// Transform functions take an arbitrary number of `serde_json::Value`to represent their arguments
+/// and return a `serde_json::Value`.
+/// the transform function itself is responsible for checking if the format and number of
+/// the arguments is correct
+///
+/// Returns a Result with an `anyhow::Error`. This allows consumers to return their own custom errors
+/// in the closure, and use `.into` to convert it into an `anyhow::Error`. The error message will be perserved
+pub type TransformFn = Box<dyn Fn(&[Value]) -> Result<Value, anyhow::Error>>;
+
+#[derive(Default)]
+pub struct Evaluator {
+    transforms: HashMap<String, TransformFn>,
 }
 
-pub fn eval_in_context<T: serde::Serialize>(input: &str, context: T) -> Result<'_, Value> {
-    let tree = Parser::parse(input)?;
-    let context = serde_json::to_value(context)?;
-    if !context.is_object() {
-        return Err(EvaluationError::InvalidContext);
-    }
-    eval_ast(tree, &context)
-}
-
-fn eval_ast<'a>(ast: Expression, context: &Context) -> Result<'a, Value> {
-    match ast {
-        Expression::Number(n) => Ok(value!(n)),
-        Expression::Boolean(b) => Ok(value!(b)),
-        Expression::String(s) => Ok(value!(s)),
-        Expression::Array(xs) => xs.into_iter().map(|x| eval_ast(*x, context)).collect(),
-
-        Expression::Object(items) => {
-            let mut map = serde_json::Map::with_capacity(items.len());
-            for (key, expr) in items.into_iter() {
-                if map.contains_key(&key) {
-                    return Err(EvaluationError::DuplicateObjectKey(key));
+impl Evaluator {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Adds a custom transform function
+    /// This is meant as a way to allow consumers to add their own custom functionality
+    /// to the expression language.
+    /// Note that the name added here has to match with
+    /// the name that the transform will have when it's a part of the expression statement
+    ///
+    /// # Arguments:
+    /// - `name`: The name of the transfrom
+    /// - `transform`: The actual function. A Box<dyn Fn(serde_json::Value) -> Value>
+    ///
+    /// # Example:
+    ///
+    /// ```rust
+    /// use jexl_eval::Evaluator;
+    /// use serde_json::{json as value, Value};
+    ///
+    /// let mut evaluator = Evaluator::new().with_transform("lower", Box::new(|v: &[Value]| {
+    ///    let s = v
+    ///            .first()
+    ///            .expect("Should have 1 argument!")
+    ///            .as_str()
+    ///            .expect("Should be a string!");
+    ///       Ok(value!(s.to_lowercase()))
+    ///  }));
+    ///
+    /// assert_eq!(evaluator.eval("'JOHN DOe'|lower").unwrap(), value!("john doe"))
+    /// ```
+    pub fn with_transform(mut self, name: &str, transform: TransformFn) -> Self {
```

Comment:
Using a builder pattern here because I think it's cool. Happy to have setters instead thou